src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableEntityContext.cs

```diff
@@ -169,58 +169,86 @@ TState IDurableEntityContext.GetState<TState>(Func<TState> initializer)
+                }
+                catch (Exception e)
+                {
+                    throw new EntitySchedulerException($"Failed to deserialize entity state: {e.Message}", e);
+                }
+            }
+            else
+            {
                 if (initializer != null)
                 {
                     try
                     {
-                        defaultValue = initializer();
+                        result = initializer();
                     }
                     catch (Exception e)
                     {
                         throw new EntitySchedulerException($"Failed to initialize entity state: {e.Message}", e);
                     }
                 }
-
-                if (this.State.EntityState != null)
-                {
-                    this.CurrentState = this.GetPopulatedState<TState>(defaultValue, initializer != null);
-                }
                 else
                 {
-                    this.CurrentState = defaultValue;
+                    result = default(TState);
                 }
-
-                this.StateWasAccessed = true;
-                this.State.EntityExists = true;
             }
 
-            return (TState)this.CurrentState;
+            this.CurrentState = result;
+            this.StateWasAccessed = true;
+            this.State.EntityExists = true;
+            return result;
         }
 
-        private TState GetPopulatedState<TState>(TState initialValue, bool usedTypeInitializer)
+        internal TState GetStateWithInjectedDependencies<TState>(Func<TState> constructor)
+            where TState : class
         {
+            this.ThrowIfInvalidAccess();
+
+            if (this.StateWasAccessed)
+            {
+                return (TState)this.CurrentState;
+            }
+
+            TState result;
+
             try
             {
-                if (usedTypeInitializer)
+                result = constructor();
+            }
+            catch (Exception e)
+            {
+                throw new EntitySchedulerException($"Failed to construct entity state object: {e.Message}", e);
+            }
+
+            if (this.State.EntityExists)
+            {
+                try
                 {
-                    // Only populate serialized state, as some fields may be populated by the initializer
-                    // using dependency injection.
-                    JsonConvert.PopulateObject(this.State.EntityState, initialValue);
-                    return initialValue;
+                    JsonConvert.PopulateObject(this.State.EntityState, result);
                 }
-                else
+                catch (Exception e)
                 {
-                    return JsonConvert.DeserializeObject<TState>(this.State.EntityState);
+                    throw new EntitySchedulerException($"Failed to populate entity state from JSON: {e.Message}", e);
```

Comment:
Would it be useful to include the type we tried to serialize into in this exception message, or is `TState` included in the stack trace we emit?

Reply:
The error messages usually contain the name of the failed function which is the same as the name of the type (when using the class-based syntax), so I think that is good enough.